block.py

```diff
@@ -1,34 +1,34 @@
-import hashlib
+from typing import List
 
-from merkle_tree import MerkleTree
-from ipv8.messaging.serialization import default_serializer
 from ipv8.messaging.lazy_payload import VariablePayload
-from ipv8.messaging.serialization import default_serializer
+
+from merkle_tree import MerkleTree
+from transaction import Transaction
+
 
 class Block(VariablePayload):
     format_list = ['varlenHutf8', 'varlenHutf8', 'varlenH-list']
     names = ['previous_hash', 'merkle_hash', 'transactions']
 
-    def __init__(self, previous_hash: str = '', merkle_hash: str = '', transactions = []):
-        self.previous_hash = str(previous_hash)
-        self.transactions = transactions if len(transactions) > 0 else []
+    def __init__(self, previous_hash: str = '', merkle_hash: str = '', transactions: List[bytes] = []):
+        # super().__init__()
+        self.previous_hash = previous_hash
+        self.transactions = transactions
         self.merkle_tree = MerkleTree()
-        self.merkle_hash = merkle_hash if merkle_hash != '' else ''
+        self.merkle_hash = merkle_hash
 
-    def add_transaction(self, transaction):
+    def add_transaction(self, transaction: Transaction):
         # We serialize the transaction before appending it to the array
         # because there seems to be an ipv8 problem when it tries to serialize the whole array
-        serialized_transaction = default_serializer.pack_serializable(transaction)
-        self.transactions.append(serialized_transaction)
-        transaction_hash = hashlib.sha256(serialized_transaction).hexdigest()
-        self.merkle_tree.add_leaf(transaction_hash)
+        self.transactions.append(transaction.get_tx_bytes())
+        self.merkle_tree.add_leaf(transaction.get_tx_hash())
```

Comment:
I run into an error when we try to call update_tree of a block. When we try to encode the child's hash to get the node_hash, it trows an error because now the hash of each child is in bytes. Before we used hexdigest instead of digest, maybe we could create a new function (get_tx_string_hash for example) that returns a string and we use that function instead of get_tx_hash.